### DISCUSSION TITLE
Official Mod launcher support - Requests

### DISCUSSION THREAD

**Discussion by Zedb0T** [open-goal/launcher#452] [Ideas]
Hi,

These are a few features I would like to see supported in the official launcher when the time comes to add mod support there.

- [ ] Be able to add a custom "list" containing information about mods/where to find them etc, so any user can host their own "list" of mods.

- [ ] Somehow find if a mod is on both lists, I guess we use the URL as a source of truth?

- [ ] Cache the latest version info and if its over a certain amount of time old or doesnt exist, then grab updated information. We are going to have to be smart and pull tricks/be efficient to avoid hitting the github API limit

- [ ] Add a clear easy to find button to delete the cache before it "expires" (Think we see a bug on some mod on twitch, and push a fix. The streamer would need to clear the cache to instantly see the corrected/patched version)

- [ ] Support simultaneous installs, at the moment you can only install one copy of Jak 1, but the unofficial mod launcher you can instantly jump between Jak the chicken, orb hunt, and Kuitars thing.

- [ ] When a mod is selected, fine out what versions are available and populate a drop down list, the top option should be default and just be "latest version" If the game is launched with latest version selected then check for update on launch and keep it updated - if a specific version is requested of course dont do this. This is pretty important if we want to support using the vanilla launcher to speedrun mods as maybe its not always "best" to run the newest version or something.

- [ ] Allow hot replacing of goal_src during the post download before compiling. See -> https://github.com/OpenGOAL-Mods/OG-ModLauncher/blob/6999ad151903ffe5ae7625b263edce30ef8b8e4b/utils/launcherUtils.py#L374C5-L405C6 . By doing it this way, If the modder leaves the value at the original default, then the launcher automatically makes the RPC nice UwU. 

- [ ] Allow the mod json to specify some kind of theme, It would be cool to see the background and maybe buttons dynamically change as you pick different mods, and would also be a clear way to see what mod is "selected" at a glance

- [ ] Maybe time to talk about "binary repos" and sharing them across mods that use the same binaries instead of having 17 copies of the exact same `gk.exe`? This is probably better left for another day but worth mentioning here

- [ ] Add some basic tracking to mods - number of installs time played progress etc and maybe push it somewhere so we can see what mods are most popular etc.

- [ ] Discuss/add some way to highlight that a mod has been released/added/new etc.

- [ ] Support some kind of flag for Mods that arent supported on the mod launcher, they show up in the list but we check this flag prompt the user that this mod/tool is not compatible with the launcher and they can download it at %url% with a disclaimer/link button

- [ ] Add some kind of disclaimer like: https://media.discordapp.net/attachments/877636980912046150/1243706296809816094/warning.png?ex=66527340&is=665121c0&hm=a5f854a67b221bc0a67885dca1197002453e6c3610c7386f7a9c9de62af996c4&=&format=webp&quality=lossless

Will update this as more ideas come to mind - Feel free to add your own post with ideas below, but limit yourself to one post you edit and keep up to date please.

**Comment by xTVaser** [comment]
- yeah
- yeah
- information will be cached in the source, not a launcher concern, no rate limit concerns either
- see previous
- yeah, though i don't think we need concurrent installs of the same mod (ie. different versions)
- yeah, you'll be able to pick the version
- this isn't really hot replacing.  Some things on this:
  - independent mods definitely do need their own unique settings, I don't like how its being done in the mod launcher -- it should be a proper first class feature from `jak-project`
  - there is no need to find-and-replace text in the config files, the decompiler can and should support json overrides via CLI flags
- yeah
- it could be done but the binaries are so small that it's not really a big concern imo (like 20mb if you exclude `goalc`).
- this can be tracked by the mod source as well, pushing up metrics is a potential future idea.

**Comment by xTVaser** [comment]
See #530 

Responding to each in a more formal way for posterity:

> Be able to add a custom "list" containing information about mods/where to find them etc, so any user can host their own "list" of mods.

Yes added, you can add "mod sources" which are just external links to JSON data that describes what mods there are, how they are supported, and where to get them from

> Somehow find if a mod is on both lists, I guess we use the URL as a source of truth?

Mods are partitioned by their source, this uses the URL as a source of truth settings-wise.  However the name of the mod source (defined within it) is used across the UX.  There is the chance of a collision, currently this is prevented at the time the user adds the source.
- In reality, source authors should avoid colliding and there likely won't be many sources -- this can be iterated on later (allow the user to override the name if there is a collision for example?)

> Cache the latest version info and if its over a certain amount of time old or doesnt exist, then grab updated information. We are going to have to be smart and pull tricks/be efficient to avoid hitting the github API limit

No rate-limit hit, the point of the architecture is that all of the necessary info is cached in the mod source.  The launcher can hit this as many times as it likes.

> Add a clear easy to find button to delete the cache before it "expires" (Think we see a bug on some mod on twitch, and push a fix. The streamer would need to clear the cache to instantly see the corrected/patched version)

Not needed, as mentioned if the mod source is updated, we can poll it everytime the relevant launcher pages are opened.  Mod source authors just need to make sure their data is up-to-date as often as they desire.

> Support simultaneous installs, at the moment you can only install one copy of Jak 1, but the unofficial mod launcher you can instantly jump between Jak the chicken, orb hunt, and Kuitars thing.

Supported, but no simultaneous installs of the same mod, but with different versions.  This could be done but felt unnecessary for a first pass.

> When a mod is selected, fine out what versions are available and populate a drop down list, the top option should be default and just be "latest version" If the game is launched with latest version selected then check for update on launch and keep it updated - if a specific version is requested of course dont do this. This is pretty important if we want to support using the vanilla launcher to speedrun mods as maybe its not always "best" to run the newest version or something.

When a mod is initially selected from the library page, it installs the latest.  From there, users can explicitly pick which version they want to use.  They are notified if an update is available.

> Allow hot replacing of goal_src during the post download before compiling. 

No! not need as per - https://github.com/open-goal/jak-project/pull/3477

> Allow the mod json to specify some kind of theme, It would be cool to see the background and maybe buttons dynamically change as you pick different mods, and would also be a clear way to see what mod is "selected" at a glance

Done, mods can provide a background image (called a cover art image) or a thumbnail image.

> Maybe time to talk about "binary repos" and sharing them across mods that use the same binaries instead of having 17 copies of the exact same gk.exe? This is probably better left for another day but worth mentioning here

Agreed, but not a launcher concern.  The bundling action can be edited (if not already) to pull from a source.  So if the mod-base started publishing binaries, the bundler script could pull those instead of grabbing from the local FS.

> Add some basic tracking to mods - number of installs time played progress etc and maybe push it somewhere so we can see what mods are most popular etc.

This is not done yet in the UI, but this is primarily a mod-source concern.  When the mod-source is grabbing all it's info (from example github releases) it can tally up the download counts and publish that info.  We should do this before finalizing the mod-source schema.

> Discuss/add some way to highlight that a mod has been released/added/new etc.

Future UX work -- my suggestion, maybe just highlight something that has had it's first release within the last month.  That way there is no state to track and will probably cover 99% of users who are checking back often.

> Support some kind of flag for Mods that arent supported on the mod launcher, they show up in the list but we check this flag prompt the user that this mod/tool is not compatible with the launcher and they can download it at %url% with a disclaimer/link button

Done, mods can be marked as external only with a URL that we open.

> Add some kind of disclaimer like

Done
